Enhancement/refactor components

swashcap opened the pull request:
Asana task:
### Goal

Make logon widget’s internals more maintainable, easier to understand.
### Methods

Convert internal components to [Matt-Esch/virtual-dom](https://github.com/Matt-Esch/virtual-dom), a React-like library.
### Testing

Run `grunt test`.
### Description

I’m already running in to mysterious bugs and difficulty adding features with this project. I believe the major problem lies in the custom code used to output DOM and maintain state. I attempted to make the widget React-like by including “components” classes, each with their own internal “state.” After playing around with React and Redux it became clear that this wasn’t necessarily the best way to do things: many internal states make things confusing. I’ve been intending to explore [Matt-Esch/virtual-dom](https://github.com/Matt-Esch/virtual-dom) after seeing [Nolan Lawson use it on pokedex.org](http://www.pocketjavascript.com/blog/2015/11/23/introducing-pokedex-org). It’s basically a light React without JSX, and could be a viable solution for future widgets we need to keep light and portable.

Here’s the component diagram via the _scripts_ directory:

```
+-- coins-logon-widget.js              - root constructor, exposes public methods
    +-- components/form.js             - form element, off which all
        +-- components/status.js       - logged in status
        +-- components/notification.js - log in error notifications
        +-- components/form-group.js   - input fields, associated errors
        +-- components/button-group.js - all form buttons
        +-- components/indicator.js    - loading spinner
```

All state is now contained in `CoinsLogonWidget` (in `CoinsLogonWidget._state`), which should make reasoning about things easier. Everything in _scripts/components/_ behaves like a [stateless React component](https://facebook.github.io/react/blog/2015/10/07/react-v0.14.html#stateless-functional-components): properties are passed down. The `CoinsLogonWidget` instance passes it’s state directly to the form component, which renders sub-components depending on the props.
### Benefits
- Easier to reason about state: it’s all in the `CoinsLogonWidget` instance
- Easier to maintain: adding/modifying code should be easier as we’re using a well-understood rendering library
- Easier to test: partially a side effect of reasonable state, but now that diffing/patching/rendering is offloaded to virtual-dom we should be able to easily test each component.
### Bonus Features

There’s a couple of additional “modes” based on how we’re actually using the logon widget:
- Redirect mode: make the widget redirect upon log in. We need this on the PR site and in CAS. Currently it requires some confusing event configuration.
- Check authorization mode: check whether the user’s logged in and display the “logged in” state. We need this on the PR site.

swashcap committed:
Move assert helpers to utils module.

swashcap committed:
Change widget to use virtual-dom for rendering.

swashcap commented:
The components have decent doc blocks. I’d like to auto-commit JSDoc documentation like [half-penny](https://github.com/MRN-Code/halfpenny) at some point so internals are easy to understand.

- [ ] Add documentation for redirect
- [ ] Add documentation for auth check mode

swashcap commented on scripts/coins-logon-widget.js:
```diff
@@ -1,115 +1,79 @@
+var patch = require('virtual-dom/patch');
+var uniqueId = require('lodash/utility/uniqueId');
 var utils = require('./lib/utils');
-require('./lib/polyfills.js');
```
Using lodash’s `bind` to minimize external deps. Although…eh, it doesn’t really matter. We can drop this when PhantomJS 2.0.0 ships to NPM.

cdaringe replied:
yea, no worries

swashcap replied:
OMG, I’m an idiot. Let’s just move the `.bind` polyfill to the test code, just like Promise.

cdaringe replied:
I plan on switching to PhantomJS full time ;)  jk jk

swashcap commented on scripts/coins-logon-widget.js:
```diff
@@ -1,115 +1,79 @@
 var utils = require('./lib/utils');
-require('./lib/polyfills.js');
-
-var EVENTS = {
```
Events now hang off `CoinsLogonWidget`, making them available for external use if needed.

swashcap commented on scripts/coins-logon-widget.js:
```diff
@@ -239,74 +255,76 @@ CoinsLogonWidget.prototype.onError = function(error) {
-        type: 'password'
-    }],
-    hiddenLabels: false,
-    horizontal: false,
```
Bye, bye, meaningless defaults. Although I just remembered the `horizontal` argument isn’t honored anymore. Crap.

swashcap committed:
Correct property case.

virtual-dom requires these to be lower case for some reason.

swashcap commented on scripts/components/button-group.js:
```diff
@@ -0,0 +1,54 @@
+ * @param {string} [props.type=button]
+ * @returns {VNode}
+ */
+function button(props) {
```
`button-group` is used to wrap buttons, and `button` should never be used directly. Perhaps move it into its own module to improve clarity.

cdaringe replied:
meh, it's probably fine in here!  maybe a note or `_` prefix?  whatevs.  call the ball

swashcap commented on scripts/lib/auth.js:
```diff
@@ -1,5 +1,5 @@
 'use strict';
-var assign = require('lodash/object/assign');
+var merge = require('lodash/object/merge');
```
Essentially the same thing. Meant to cut down on package size as `assign` isn’t used, but `merge` is now used in _scripts/coins-logon-widget.js_.

cdaringe replied:
oh man!  i didn't even know merge was a thing!  so cool.

swashcap commented on test/index.js:
```diff
@@ -132,13 +132,14 @@ test.only('initial logged in state', function(t) {
     // Hijack widget internals to ensure checks fire at the appropriate time
     // TODO: Figure out better way to do this, maybe spies
-    var originalSetToLogout = myWidget.form.setToLogout;
-    myWidget.form.setToLogout = function(text) {
```
No more private API calls in the test. :+1:

swashcap committed:
Add documentation for initial auth check mode.

swashcap committed:
Add redirect mode documentation

swashcap committed:
Drop widget version property.

Webpack injects the version into dist/, but any code `require`-ing the
widget's source code errors because `__VERSION__` still exists in the
source. Perhaps SED or the like can replace a placeholder and automatically
commit the change.

swashcap committed:
Remove test isolate running method.

This makes only the 'initial logged in state' test run. Yikes\!

swashcap committed:
Add auth storage clearing to test.

This prevents async auth storage mutation in other tests from messing with
the auth storage state for this test.

cdaringe commented on scripts/coins-logon-widget.js:
```diff
@@ -1,115 +1,79 @@
+var bind = require('lodash/function/bind');
+var createElement = require('virtual-dom/create-element');
+var diff = require('virtual-dom/diff');
+var EventEmitter = require('wolfy87-eventemitter');
```
what's the scoop with this little guy?

swashcap replied:
It's a browser-ready event emitter library which `CoinsLogonWidget` extends. We could just bundle Node's `EventEmitter`, too.

cdaringe commented on scripts/coins-logon-widget.js:
```diff
@@ -121,37 +85,83 @@ CoinsLogonWidget.prototype._setState = function() {
+    delete this._options;
+    delete this._rootNode;
+    delete this._state;
+    delete this._tree;
```
if the instance is going to be de-referenced and garbage collected, probably no need to scrap all the guts?

swashcap replied:
I’m actually not sure how this happens. Would calling `myWidget.destroy()` automatically GC all this? Maybe I could test the soft logon widget – I have a feeling a reference is unintentionally maintained.

cdaringe replied:
nah, it would be GC'd when whatever instantiates `myWidget` is no longer referenced by anything in memory.  so, a page change, obviously.  or `delete myWidget`.

cdaringe commented on scripts/coins-logon-widget.js:
```diff
@@ -185,49 +196,54 @@ CoinsLogonWidget.prototype.login = function(formData) {
+    var patches = diff(this._tree, newTree);
+
+    this._rootNode = patch(this._rootNode, patches);
+    this._tree = newTree;
```
cool.

cdaringe commented:
this looks SPECTACULAR.  truly excellent work, per usual, @swashcap!

swashcap committed:
Refactor so tests pass.

This addresses issues with the refactored `CoinsLogonWidget` that prevented
test code from running. Some test code was adjusted to reflect the new
components, too.

swashcap committed:
Ensure `horizontal` option is honored.

swashcap committed:
Fix notification's text display, add notification tests.

swashcap committed:
Add tests for custom messages.

swashcap committed:
Move `bind` to a polyfill in the test.

This removes lodash's `bind` from the bundle and polyfills PhantomJS, the
only environment that needs it.

swashcap committed:
Make 'button' component more obviously private.

cdaringe commented:
i guess what i'm sayin is who's wolfy?  do you trust the wolf man?  it struck me as `require('homelessman62-react')` ;)

swashcap committed:
Add redirect button feature.

This ensures a redirect button is present when the widget is in the
'logged in' state. Associated tests are added, too.

swashcap committed:
Remove unneeded `delete`s.

swashcap commented:
> `require('homelessman62-react')` ;)

Ha. When I was looking for UMD-wrapped event emitters it was the best option. [1.2k downloads on NPM](https://www.npmjs.com/package/wolfy87-eventemitter), and the docs are good. I think “wolfy87” is legacy, either the project changed ownership or the guy changed his name to “olical”.

swashcap commented:
Sorry for all the CI checks. I `npm link`-ed and checked the PR site, CAS and soft login widget. It works! We’re free to upgrade at our leisure. I can open those PRs.
